webapp-django/crashstats/topcrashers/views.py

```diff
@@ -0,0 +1,302 @@
+
+            # Number of hang crashes.
+            hit['hang_count'] = 0
+
+            sig_hang = hit['facets']['hang_type']
+            for row in sig_hang:
+                if row['term'] in (1, -1):
+                    hit['hang_count'] = row['count']
+
+        # Run the same query but for the previous date range, so we can
+        # compare the rankings and show rank changes.
+        dates = get_date_boundaries(params)
+        delta = (dates[1] - dates[0]) * 2
+        params['date'] = [
+            '>=' + (dates[1] - delta).isoformat(),
+            '<=' + dates[0].isoformat()
+        ]
+
+        previous_range_results = api.get(**params)
+        total = previous_range_results['total']
+
+        compare_signatures = {}
+        if (total > 0 and 'signature' in previous_range_results['facets']):
+            signatures = previous_range_results['facets']['signature']
+            for i, hit in enumerate(signatures):
+                compare_signatures[hit['term']] = {
+                    'count': hit['count'],
+                    'rank': i + 1,
+                    'percent': 1.0 * hit['count'] / total
+                }
+
+        for hit in results:
+            sig = compare_signatures.get(hit['term'])
+            if sig:
+                hit['diff'] = hit['percent'] - sig['percent']
+                hit['rank_diff'] = hit['rank'] - sig['rank']
+                hit['previous_percent'] = sig['percent']
+            else:
+                hit['diff'] = 'new'
+                hit['rank_diff'] = 0
+                hit['previous_percent'] = 0
+
+    return search_results
+
+
+@waffle_switch('topcrashers-real-time')
+@pass_default_context
+@anonymous_csrf
+@check_days_parameter([1, 3, 7, 14, 28], default=7)
+def topcrashers(request, days=None, possible_days=None, default_context=None):
+    context = default_context or {}
+
+    product = request.GET.get('product')
+    versions = request.GET.get('version')
+    crash_type = request.GET.get('process_type')
+    os_name = request.GET.get('platform')
+    result_count = request.GET.get('_facets_size')
+
+    if product not in context['releases']:
+        raise http.Http404('Unrecognized product')
+
+    if not versions:
+        # :(
+        # simulate what the nav.js does which is to take the latest version
+        # for this product.
+        for release in context['currentversions']:
+            if release['product'] == product and release['featured']:
+                url = '%s&version=%s' % (
+                    request.build_absolute_uri(), urlquote(release['version'])
+                )
+                return redirect(url)
+    else:
+        versions = versions.split(';')
+
+    if len(versions) == 1:
+        context['version'] = versions[0]
+
+    release_versions = [x['version'] for x in context['releases'][product]]
+    if context['version'] not in release_versions:
+        raise http.Http404('Unrecognized version')
```

Comment:
Ugh! I sadly recognize so much of these 20 lines above. It hurts my eyes every time. 
SOME day we'll get this simplified and beautiful again.